### PR TITLE
App create command in monorepo setup

### PR DESCRIPTION
**Pull request by lukasz-kaniowski**
Assuming `lukasz-app-3` is an existing repository we can add a new application like this:

```shell
cd lukasz-app-3
corectl app create lukasz-sub-app --tenant lukasz-sub-app --from-template blank
Added lukasz-sub-app to repository: https://github.com/coreeng/lukasz-app-3
                                                                                                                                       
Note: to complete application onboarding to the Core Platform you have to first merge PR that adds this application to your repository.
PR url: https://github.com/coreeng/lukasz-app-3/pull/5                                                                                 
                                                                                                                                       
After the PR is merged, your application is ready to be deployed to the Core Platform!                                                 
It will either happen with next commit or you can do it manually by triggering P2P workflow.                                           
To do it, use GitHub web-interface or GitHub CLI.                                                                                      
Workflows link: https://github.com/coreeng/lukasz-app-3/actions                                                                        
GitHub CLI commands:                                                                                                                   
  gh workflow list -R coreeng/lukasz-app-3                                                                                             
  gh workflow run <workflow-id> -R coreeng/lukasz-app-3             
```

I've added a longer help description

```shell
corectl app create -h           
Creates new application:

- create a git repository locally and initializes it with application skeleton generated from a template
- template is rendered with arguments: {{name}}, {{tenant}}
- create a new github repository with p2p related variables
- create a new PR to environment repository with configuration update for the new application (if necessary)

NOTE: If parent directory is an existing git repository it will add a new application to it 
      without creating a new remote repository.

Usage:
  corectl app create <app-name> [<local-path>] [flags]
...
```

I'm passing `tenant` and `name` to render templates so they can be used in software templates.

### CODE REVIEW

File: doc/adr/0003-app-create-monorepo.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
+# p2p promote command
```
Comment (tombart) [inline doc/adr/0003-app-create-monorepo.md:1]: title needs updating

File: doc/adr/0003-app-create-monorepo.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+Date: 2024-08-07
+
+## Status
+
+Accepted
```
Comment (tombart) [inline doc/adr/0003-app-create-monorepo.md:7]: proposed?

File: doc/adr/0003-app-create-monorepo.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+Changes to current implementation
+
+1. if local-path is already a git repository
+   - don't create a new github repository
+   - don't set github variables (as they already set by parent)
+   - check if there is no local changes
+2. `tenant` should be passed to template to be rendered in Makefile
+3. `.github` folder from templates would have to be copied to root of the repository
+and all files would be prefixed with `app-name`
```
Comment (tombart) [inline doc/adr/0003-app-create-monorepo.md:80]: are there any additional changes to p2p? the design is we'll have separate pipeline per repo within parent repo, correct?

File: doc/adr/0003-app-create-monorepo.md (hunk 2 above)
Comment (lukasz-kaniowski) [reply doc/adr/0003-app-create-monorepo.md:7]: it's always accepted when merge, right?

File: doc/adr/0003-app-create-monorepo.md (hunk 3 above)
Comment (lukasz-kaniowski) [reply doc/adr/0003-app-create-monorepo.md:80]: yes, separate pipeline within parent repo https://github.com/coreeng/developer-platform/blob/main/doc/architecture/decisions/0057-reference-app-monorepo.md#pipelines

there will be some changes to Makefiles in reference templates

File: pkg/application/create.go [hunk 4]
```diff
@@ -76,29 +201,53 @@ func Create(op CreateOp, githubClient *github.Client) (result CreateResult, err
...
+	err = os.RemoveAll(filepath.Join(path, ".github"))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func githubWorkflowsExist(path string) bool {
+	githubWorkflowsPath := filepath.Join(path, ".github", "workflows")
+	dir, err := os.ReadDir(githubWorkflowsPath)
```
Comment (tombart) [inline pkg/application/create.go:295]: not sure, should we log the error here?

File: pkg/application/create.go [hunk 5]
```diff
@@ -122,16 +271,24 @@ func ValidateCreate(op CreateOp, githubClient *github.Client) error {
...
+	if isMonorepo {
+		_, response, err := githubClient.Repositories.Get(
+			context.Background(),
+			op.OrgName,
+			op.Name,
+		)
+		if err == nil {
+			return fmt.Errorf("%s/%s repository already exists", op.OrgName, op.Name)
+		}
+		if err != nil && response.StatusCode != http.StatusNotFound {
```
Comment (tombart) [inline pkg/application/create.go:335]: probably don't need err != nil here, it'll always be not nil if it gets here

File: pkg/application/create.go [hunk 6]
```diff
@@ -30,41 +33,209 @@ type CreateOp struct {
...
+		panic(undo.FormatError("create new application", fmt.Errorf("%v", err), errs))
+	}
+}
+
+func prepareLocalPath(localPath string, undoSteps *undo.Steps) error {
+	if err := os.MkdirAll(localPath, 0o755); err != nil {
+		return err
+	}
+	undoSteps.Add(func() error {
+		return os.RemoveAll(localPath)
```
Comment (tombart) [inline pkg/application/create.go:186]: not sure if that's the case, but worth checking, if I create monorepo, this undo step won't delete my existing parent dir? just the submodule within that project?

File: pkg/application/create_test.go [hunk 7]
```diff
@@ -255,14 +265,184 @@ var _ = Describe("Create new application", func() {
...
+		It("returns empty PR url", func() {
+			Expect(createResult.PRUrl).To(Equal(""))
+		})
+
+		It("returns false for monorepo mode", func() {
+			Expect(createResult.MonorepoMode).To(BeFalse())
+		})
+
+	})
+	Context("monorepo mode", Ordered, func() {
```
Comment (tombart) [inline pkg/application/create_test.go:291]: maybe would be good to add one test that fails to create the repo, testing undo operations (like your github mock failing when creating pr or something like this)

File: pkg/application/create_test.go [hunk 8]
```diff
@@ -13,6 +13,7 @@ import (
 	"github.com/migueleliasweb/go-github-mock/src/mock"
```
Comment (tombart) [inline pkg/application/create_test.go:13]: also could we add one integration test? to prove it works from outside (as blackbox), should be similar to one we already have for app create

File: pkg/application/create.go (hunk 4 above)
Comment (lukasz-kaniowski) [reply pkg/application/create.go:299]: added better error handling

File: pkg/application/create.go (hunk 5 above)
Comment (lukasz-kaniowski) [reply pkg/application/create.go:345]: haha, not sure how this get here

File: pkg/application/create.go [hunk 9]
```diff
@@ -30,41 +33,209 @@ type CreateOp struct {
...
 	if err != nil {
 		return result, err
 	}
-	undoSteps.Add(func() error {
-		return os.RemoveAll(op.LocalPath)
-	})
-	if op.Template != nil {
-		if err = template.Render(op.Template, op.LocalPath); err != nil {
+
+	if isMonorepo {
```
[the remark this answers is not in the document]
Comment (lukasz-kaniowski) [reply pkg/application/create.go:60]: done

File: pkg/application/create.go [hunk 10]
```diff
@@ -30,41 +33,209 @@ type CreateOp struct {
...
+
+	return localRepo, isMonorepo, nil
+}
+
+func openMonorepoMaybe(localPath string) (localRepo *git.LocalRepository, isMonorepo bool, err error) {
+	localRepo, err = git.OpenLocalRepository(filepath.Dir(localPath))
+	if err != nil && !errors.Is(err, gogit.ErrRepositoryNotExists) {
+		return nil, false, err
+	}
+	isMonorepo = localRepo.Repository() != nil
```
[the remark this answers is not in the document]
Comment (lukasz-kaniowski) [reply pkg/application/create.go:217]: we are getting error that is **not** ErrRepoNotExist, so we still need to check

File: pkg/application/create.go (hunk 6 above)
Comment (lukasz-kaniowski) [reply pkg/application/create.go:190]: added a test

File: pkg/application/create_test.go (hunk 7 above)
Comment (lukasz-kaniowski) [reply pkg/application/create_test.go:292]: added a test

File: pkg/application/create.go [hunk 11]
```diff
@@ -30,41 +33,209 @@ type CreateOp struct {
...
+		&github.NewPullRequest{
+			Base:  github.String(git.MainBranch),
+			Head:  github.String(branchName),
+			Title: github.String("Add " + name + " application"),
+			Body:  github.String(fmt.Sprintf("Adding `%s` application", name)),
+		})
+	return pullRequest, err
+}
+
+func getRemoteRepositoryFullId(op CreateOp, githubClient *github.Client, localRepo *git.LocalRepository) (git.GithubRepoFullId, error) {
```
[the remark this answers is not in the document]
Comment (lukasz-kaniowski) [reply pkg/application/create.go]: pointer make sense

**Review by tombart — APPROVED** [review]

**Pull request merged**